### PR TITLE
cephadm: Add `chown` to unit.run for adoped simple OSDs

### PR DESCRIPTION
**Pull request by tserong**
We're not using `ceph-volume simple activate` anymore, so need to add `chown ceph:ceph` for each of block, block.db and block.wal to ensure old-style simple OSDs are able to start.

This PR also addresses a couple of the comments in https://github.com/ceph/ceph/pull/34565, but I didn't change it to loop over all devices when looking at ceph.type in check_offline_lvm_osd() - based on the ceph-volume lvm prepare code, there may be several different types, but finding one that's 'block' indicates bluestore, and finding one that's 'data' indicates filestore.
    
Fixes: https://tracker.ceph.com/issues/45129
Signed-off-by: Tim Serong <tserong@suse.com>

### CODE REVIEW

**Comment by tserong** [comment]
Just to make the output of this change clear, it injects e.g. the following three lines at the start of unit.run:

```
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block.db ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block.db
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block.wal ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.10/block.wal
```

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -1786,6 +1786,10 @@ def deploy_daemon_units(fsid, uid, gid, daemon_type, daemon_id, c,
         if daemon_type == 'osd':
             # osds have a pre-start step
             assert osd_fsid
+            if osd_simple:
+                for n in ['block', 'block.db', 'block.wal']:
+                    p = os.path.join(data_dir, n)
+                    f.write('[ -L {p} ] && chown {uid}:{gid} {p}\n'.format(p=p, uid=uid, gid=gid))
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:1792]:
```suggestion
                    code = """
# This is a simple OSDs. Needs chown on startup:
[ -L {p} ] && chown {uid}:{gid} {p}\n""".format(p=p, uid=uid, gid=gid)
                    f.write(code)
```
?

Comment (tserong) [reply src/cephadm/cephadm:1792]: Good idea, although that will put the comment before each of the three chown lines, e.g.:

```
# This is a simple OSDs. Needs chown on startup:
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block
# This is a simple OSDs. Needs chown on startup:
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block.db ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block.db
# This is a simple OSDs. Needs chown on startup:
[ -L /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block.wal ] && chown 167:167 /var/lib/ceph/f6cf67cc-dcfa-3de8-a33f-cab60875412f/osd.21/block.wal
```

We could alternately do:

```diff --git a/src/cephadm/cephadm b/src/cephadm/cephadm
index 0652d89b280..18d1b30eba1 100755
--- a/src/cephadm/cephadm
+++ b/src/cephadm/cephadm
@@ -1787,6 +1787,7 @@ def deploy_daemon_units(fsid, uid, gid, daemon_type, daemon_id, c,
             # osds have a pre-start step
             assert osd_fsid
             if osd_simple:
+                f.write('# This is a simple OSD. Needs chown on startup:\n')
                 for n in ['block', 'block.db', 'block.wal']:
                     p = os.path.join(data_dir, n)
                     f.write('[ -L {p} ] && chown {uid}:{gid} {p}\n'.format(p=p, uid=uid, gid=gid))
```

...which would put the comment only once at the start.

Thoughts?

Comment (sebastian-philipp) [reply src/cephadm/cephadm:1792]: much bettter!

Comment (tserong) [reply src/cephadm/cephadm:1793]: Done!

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -1786,6 +1786,11 @@ def deploy_daemon_units(fsid, uid, gid, daemon_type, daemon_id, c,
         if daemon_type == 'osd':
             # osds have a pre-start step
             assert osd_fsid
+            if osd_simple:
+                f.write('# This is a simple OSD. Needs chown on startup:\n')
+                for n in ['block', 'block.db', 'block.wal']:
+                    p = os.path.join(data_dir, n)
+                    f.write('[ -L {p} ] && chown {uid}:{gid} {p}\n'.format(p=p, uid=uid, gid=gid))
```
[the remark this answers is not in the document]
Comment (tserong) [reply src/cephadm/cephadm:1793]: Why flip the logic? Aren't the two equivalent, or am I missing some subtlety?

**Comment by tserong** [comment]
> If possible, I think it would be nice to avoid adding the `osd_simple` bool to `deploy_daemon_units` ...

Now that you mention it, it's probably safe to always have the chmods in there, as they happen first before the lvm activate step. If that's true, we can ditch the `osd_simple` bool. I'll do a little further testing.

**Comment by tserong** [comment]
Confirmed, chowning these things doesn't have any adverse affect on LVM OSDs, so we could get rid of the `osd_simple` flag. OTOH, that would mean that all OSDs will always have these chown lines, even newly created ones (i.e. not only adopted ones), which while harmless may arguably be slightly messy.

@sebastian-philipp, what do you think?

**Comment by sebastian-philipp** [comment]
Right. Not having to adjust the logic here make things simpler.

@jan--f do you have anything against unconditionally doing the chown?

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-04-24_13:37:19-rados-wip-swagner2-testing-2020-04-24-1154-distro-basic-smithi/

Failure: https://tracker.ceph.com/issues/45194

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-04-26_09:25:11-rados-wip-kefu-testing-2020-04-26-1313-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-26_13:55:50-rados-wip-kefu-testing-2020-04-26-1314-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-27_01:47:10-rados-wip-kefu-testing-2020-04-26-1314-distro-basic-smithi/

need to rerun the failed tests without #34728

File: src/cephadm/cephadm (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/cephadm/cephadm:1793]: it avoids a non-zero exit code when `{p}` is not a symlink

[the remark this answers is not in the document]
Comment (tserong) [reply src/cephadm/cephadm]: Gotcha, thanks!

**Comment by tserong** [comment]
OK, I've removed the osd_simple flag, tweaked the comments slightly, and flipped the `[ ! -L ] || chown` logic.

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -1777,7 +1777,7 @@ def deploy_daemon(fsid, daemon_type, daemon_id, c, uid, gid,
 
 def deploy_daemon_units(fsid, uid, gid, daemon_type, daemon_id, c,
                         enable=True, start=True,
-                        osd_fsid=None):
+                        osd_fsid=None, osd_simple=False):
     # type: (str, int, int, str, Union[int, str], CephContainer, bool, bool, Optional[str]) -> None
```
[the remark this answers is not in the document]
Comment (tserong) [reply src/cephadm/cephadm]: I've ditched the osd_simple flag now.

**Comment by sebastian-philipp** [comment]
image

any update on this run?

**Comment by mgfritch** [comment]
http://pulpito.ceph.com/mgfritch-2020-05-05_02:07:06-rados-wip-mgfritch-testing-2020-05-04-1622-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
> http://pulpito.ceph.com/mgfritch-2020-05-05_02:07:06-rados-wip-mgfritch-testing-2020-05-04-1622-distro-basic-smithi/

https://tracker.ceph.com/issues/45335

**Pull request merged**